### PR TITLE
Added Rails 5.0 support

### LINKED ISSUE #249: Compatibility with Rails 5.0
With the release of Rails 5.0.0.beta1 just days ago, many users (myself included) are seeking to upgrade their current Rails 4.2 apps to ensure forward-compatibility with this new version of Rails and to make use of its new features and capabilities.

Currently, the `audited.gemspec` includes the line:
```
gem.add_dependency 'activerecord', '~> 4.0'
```
which prevents usage of this gem (or even experimentation with it) in this new version of Rails.

I would like to request that audited be made compatible with Rails 5.0.

### PR DESCRIPTION
**Pull request by mvastola**
Resolves #249.

**Note**: As `rails-observers` has not yet released a version with Rails 5.0 compatibility (the `appraisals` `Gemfile` in this PR points to a commit in `rails-observers` Github), but can still be merged without issue. To be used with Rails 5.0, any application containing a version of `audited` in its `Gemfile` with this PR merged will also need the following entry:
```
gem 'rails-observers', :github => 'rails/rails-observers', :branch => 'master'
```

This `rails50` appraisal in this repository should be updated to point to `rails-observers` on RubyGems once a version is released that is compatible with Rails 5.0.

### CODE REVIEW

**Comment by mvastola** [comment]
@danielmorrison Please note that, according it its [`README`](https://github.com/rails/protected_attributes/blob/master/README.md), `protected_attributes` is no longer maintained as of the time of the release of Rails 5, so I have updated the test suite with the assumption that -- at the very least -- `audited` will be not support this feature for Rails 5 apps. (This gem also broke functionality when it was enabled under Rails 5 in ways it shouldn't have.)

A question that the owners of this gem should answer, however, is whether `audited` should abandon all support for `protected_attributes` whatsoever, since it is soon to be abandonware.

**Comment by mvastola** [comment]
All tests are now passing and this will be ready to merge pending a release of `rails-observers`.

**Comment by mvastola** [comment]
@danielmorrison Could you also clarify the current support status for Rails 4.0 and 4.1? The `README.md` currently says:

> Audited currently (4.x) works with Rails 4.2. For Rails 3, use gem version 3.0 or see the [3.0-stable branch](https://github.com/collectiveidea/audited/tree/3.0-stable).

But is silent on 4.0 and 4.1, which are still tested by Travis CI in the master branch. I took care to make sure the tests for 4.0 and 4.1 continue to pass when making my edits, but I'm not sure if you plan on abandoning support for these versions when 5.0 is released, have already done so, or only quasi-support them. In any case I think it's probably advisable to state the current policy in `README.md` (as well as the future policy, if it is known and will change). 

If `audited` is no longer supporting 4.0 and/or 4.1, please let me know if I should be removing those tests from `.travis.yml`.

For now I changed the `README.md` to state that the gem works in 4.2 and 5.0 and that the gem may work in 4.0 and 4.1 but that functionality in the later two isn't guaranteed. I consider this change totally tentative though, pending comment from `audited`'s owners.

**Comment by mvastola** [comment]
**Update**: I'm marking this PR as ready to merge since there are no known regressions and `audited`'s runtime dependencies cause it to obtain only stable releases of `rails-observers` from RubyGems. Therefore, merging this request (and even pushing a new release of this gem) should have no adverse effects whatsoever. 

Until a new release of `rails-observers` is pushed to RubyGems, `audited` will continue to not work on Rails 5.0, with the exception that advanced users would be able to ensure functionality if they manually added the following entry to their `Gemfile`s:
```
gem 'rails-observers', :github => 'rails/rails-observers', :branch => 'master'
```

**Comment by Azzurrio** [comment]
@mvastola I've an issue with this, `controller` method inside `lib/audited/sweeper.rb` is always returns `nil`, which means that fields like `user`, `remote_address` & `request_uuid` is never been set and always returns `nil`. Does this related to the new version or something else?

**Comment by mvastola** [comment]
@Azzurrio Unlear.. I would need to know more. Is your repository public somewhere for me to look at? 
Otherwise, are you sure you're checking the `controller` method *during* the action? Can you send me a gist demonstrating what you think is broken?

**Comment by Azzurrio** [comment]
@mvastola I've a simple update action, something just like `product.update(product_params)`. The auditing is working perfectly, so any change I've made to the object works normally except the fields I've just mentioned above. I've added a `binding.pry` inside update action and I can access devise `current_user` method normally. Then I've open the gem using `bundle open audited` and I put another `binding.pry` inside `before_create` method inside `lib/audited/sweeper.rb` like this:
```
def before_create(audit)
      binding.pry
      audit.user ||= current_user
      audit.remote_address = controller.try(:request).try(:remote_ip)
      audit.request_uuid = request_uuid if request_uuid
    end
```
I retried the request again and when binding stops here I tried to access `current_user` & `controller`. They always return `nil`

**Comment by mvastola** [comment]
@Azzurrio Odd. Well I just pushed a change (deleted a line of code that wasn't mine but shouldn't have been there nevertheless), so try that. If that doesn't work, try doing modifying `Audited::Sweeper.around` to add a `binding.pry` before and after the `yield`. I'm curious to know if `Audited::Sweeper.around` is ever actually invoked and if your `before_create` takes place inside the `yield` as it should.

**Comment by mvastola** [comment]
@Azzurrio I want to make sure I'm duplicating your setup on my end when I try to simulate your bug. There was never an alpha release of Rails 5.0 released to RubyGems. I'm guessing you're just using a particular commit from the `rails/rails` github for your project? Can you give me the sha1 of the commit you are using?

**Comment by Azzurrio** [comment]
@mvastola Nothing changed. And yes! I've added the `binding` before and after the `yield` and It's never invoked.

**Comment by Azzurrio** [comment]
I am using rails like that
```
gem "rails", github: "rails/rails", ref: "7db7b287ecd616d61dd09147888d02b74d219dd1"
```

**Comment by Azzurrio** [comment]
Also I'm serving and API. so I just required this in my `application.rb` file
```
require "rails"
require "active_model/railtie"
require "active_job/railtie"
require "active_record/railtie"
require "action_controller/railtie"
```

**Comment by mvastola** [comment]
Ok. My bet is the problem is in your alpha version of rails, but let me check this out. Just to confirm.. this error is inside an `ActionController` (a subclass of `ActionController::Base`)?

**Comment by Azzurrio** [comment]
These are the list of superclasses
`ProductsController` < `ApplicationController` < `ActionController::API` < `ActionController::Metal` < `AbstractController::Base`

**Comment by mvastola** [comment]
**Oh!** Well that's the problem right there. `audited` isn't coded to work with APIs (and I suppose I need to code it so it is). See the last block of code in `lib/audited/sweeper.rb`:
```
if defined?(ActionController) && defined?(ActionController::Base)
  ActionController::Base.class_eval do
    around_action Audited::Sweeper.instance
  end
end
```

**Comment by mvastola** [comment]
I'll have a fix done in 5-10.

**Comment by Azzurrio** [comment]
Yeah thats it! :+1:

**Comment by mvastola** [comment]
@Azzurrio Should be all set!

**Comment by Azzurrio** [comment]
@mvastola Perfect! works now. Good Job :+1:

**Comment by mvastola** [comment]
@danielmorrison FYI, going through the changelogs, I just came across definitive proof [here](https://github.com/rails/rails/commit/f4fbc0301021f13ae05c8e941c8efc4ae351fdf9?short_path=1ed2907#diff-1ed2907b3b8f148c2533558a77673ffa) that `protected_attributes` is no longer supported in Rails 5.0.

Also, @danielmorrison, do you happen to have any ETA on reviewing of this PR? I only ask because you [expressed willingness](https://github.com/collectiveidea/audited/issues/249#issuecomment-166311840) to personally write this entire thing yourself nearly two weeks ago, so I was hoping you had time.. 

Please let me know what, if any, changes/clarifications you'd like made or any questions I can answer.

**Comment by mvastola** [comment]
@danielmorrison PS: I know I had been editing it endlessly, so apologies if you were waiting on me to stop, but I swear I'm done now (barring new, as-of-yet undiscovered issues with the PR). :-p

**Comment by mvastola** [comment]
@danielmorrison would it be possible to get this PR reviewed? 

I understand if you may be hesitant to merge it just *quite* yet because it has to point to [the master branch of rails-observers](https://github.com/rails/rails-observers) (pending the next release of that gem, which I am [following up on](https://github.com/rails/rails-observers/pull/39#issuecomment-178971871) as it was promised almost a month ago).

That said, considering the tentative dates for the RC(s) and subsequent initial release for Rails 5 are rapidly approaching (both scheduled for later in February), it would be ideal to get a jump start on any changes/revisions you might want made to this PR before you consider it worthy to be merged. Ideally, it is obviously preferable to allow as much time as possible between the introduction of Rails 5 compatibility in `audited` and the final release of Rails 5.0.0.

File: lib/audited/auditor.rb [hunk 1]
```diff
@@ -217,7 +217,8 @@ def write_audit(attrs)
       def require_comment
         if auditing_enabled && audit_comment.blank?
           errors.add(:audit_comment, "Comment required before destruction")
-          return false
+          return false if Rails.version.start_with?('4.')
+          throw :abort
```
Comment (danielmorrison) [inline lib/audited/auditor.rb]: why the `throw :abort` here?

**Comment by danielmorrison** [comment]
@mvastola great work. Sorry for the delay, my free time has been limited lately. 

I had one question, otherwise this looks good to me!

**Comment by mvastola** [comment]
From [activerecord/CHANGELOG.md](https://github.com/rails/rails/blob/099ddfdefd44fda11d0f6a72f934f8a0ee83141b/activerecord/CHANGELOG.md):

> Change the way in which callback chains can be halted.

> The preferred method to halt a callback chain from now on is to explicitly `throw(:abort)`. In the past, returning `false` in an Active Record `before_` callback had the side effect of halting the callback chain. This is not recommended anymore and, depending on the value of the `ActiveSupport.halt_callback_chains_on_return_false` option, will either not work at all or display a deprecation warning.

**Comment by danielmorrison** [comment]
Cool. I learned something new!

File: lib/audited/auditor.rb (hunk 1 above)
Comment (mvastola) [inline lib/audited/auditor.rb]: See [this comment](https://github.com/collectiveidea/audited/pull/252#issuecomment-179853559) to the PR.

**Comment by mvastola** [comment]
Thanks for reviewing. I actually hadn't meant for you to merge *just* yet since I'm running into hurdles getting a  version of `rails-observers` on Ruby Gems that supports Rails 5, but I suppose this can't hurt. I would caution against a new release of this gem at this time though.

**Comment by mvastola** [comment]
PS: I understand your time is super limited, but if you have any insight you could lend at https://github.com/rails/rails-observers/pull/41 it would be much appreciated. Barring that, a comment from you on the PR as the owner of this gem might facilitate my receiving some assistance from the maintainers (whose most recent response time is 29 days and counting).

Thanks!

**Comment by danielmorrison** [comment]
Ha, well master can be wonky for a few days. I'll take a peek later today.

**Comment by mvastola** [comment]
Thanks. The only "wonkiness" is rependencies won't resolve correctly if someone is pulling `audited` from the github `master` branch but plain vanilla RubyGems to fetch `rails-observers`, all whIle running a betask release of `rails`.

Which is to say, their app would have to inherentlt be rather wonky to start with in order for the combination of events necessary for triggering a problem to occur.

**Comment by ericboehs** [comment]
Now that Rails 5 is out, will a new release be cut? Or are we waiting on rails-observers?

**Comment by shakycode** [comment]
👍

**Comment by danielmorrison** [comment]
I plan to look into it this afternoon and see what (if anything is still missing).

**Comment by danielmorrison** [comment]
I cut a new PR #276 for a release, but `rails-observers` needs a new version cut. 

If you want to try with Rails 5 today, use: 

```
gem 'audited', github: 'collectiveidea/audited', branch: 'rails5'
gem 'rails-observers', github: 'rails/rails-observers', branch: 'master'
```

and let me know on the PR if you have any problems!

**Comment by ericboehs** [comment]
Great @danielmorrison! @rafaelfranca is planning to release a new version of rails-observers in "the next days". https://github.com/rails/rails-observers/commit/876c522184a4ed26d564833f496ef1bfa7af654a

**Comment by rafaelfranca** [comment]
:+1: `rails-observers` is on the top of my list since we use in Shopify.

**Comment by mvastola** [comment]
`rails-observers` has `rails` 5 support merged in, but there was [a minor regression](https://github.com/rails/rails-observers/issues/45) in my PR to provide that support (basically you have to manually add an extra line of code to enable each observer). I submitted a new [PR](https://github.com/rails/rails-observers/pull/41) to fix this regression a while ago, but it sat dormant for a bit, and since then merge conflicts have arisen. 

When this got bumped last night, I tried to fix the merge conflicts, but doing so in the most straightforward manner seems to have unfixed this bug. I'm working on updating the PR and then it should be good on my end for @rafaelfranca to push.

**Comment by quentindemetz** [comment]
was this ever merged?

**Pull request merged**